src/hooks/use-fetch.js

```diff
@@ -0,0 +1,55 @@
+import { useEffect, useState } from 'react';
+
+const isBadResponse = response => !response || (response && !response.ok);
+
+const getErrorMessage = (response, method) =>
+    response
+        ? `${method} Error: ${response.status} - ${response.statusText}`
+        : `${method} Error: Did not receive a response from the server`;
+
+/**
+ * Hook with option to debounce to fetch data (GET) and provide any postprocessing
+ * @param {*} url The url to fetch from
+ * @param {*} postprocessData Manipulations to data to be done after a fetch
+ * @param {*} debounceTime (Optional) if provided, will debounce the fetch based on this number of ms
+ */
+function useFetch(url, postprocessData, debounceTime) {
+    const [data, setData] = useState(null);
+    const [error, setError] = useState(null);
+    const [fetchEvent, setFetchEvent] = useState(null);
+    const fetchData = async () => {
+        const resp = await fetch(url);
+        if (isBadResponse(resp)) {
+            const errorMessage = getErrorMessage(resp, 'GET');
+            console.warn(errorMessage);
+            setError(errorMessage);
+            setData(null);
+        } else {
+            const parsedData = await postprocessData(resp);
```

Comment:
Curious about the approach here to make `postprocessData` an async function. Is this meant to do things like parse the response as json?

Reply:
Yep that is my thinking, would it make sense to just maybe return `data` in this case?

Reply:
If we are just using this to parse the response as json I wonder if we can ensure that the api's return json so this step can be skipped. That way we won't need to convert the data to json and can just return it directly.

But if there are some other async operations we want to do on the data before returning it then using this `postProcessData` approach definitely makes sense. What do you think?

Reply:
Ah I think that is a great idea, lets do the `json` here like we do in University and postprocess via another hook when this is to be used. I went ahead and made the change